nucliadb_node/src/shards/providers/shard_cache/resource_cache.rs

```diff
@@ -0,0 +1,427 @@
+// Copyright (C) 2021 Bosutech XXI S.L.
+//
+// nucliadb is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at info@nuclia.com.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::HashMap;
+use std::hash::Hash;
+use std::num::NonZeroUsize;
+use std::sync::{Arc, Condvar, Mutex, Weak};
+
+use lru::LruCache;
+
+// Classic implementation of a binary semaphore, used to be able to
+// block while an entry is being loaded by another thread, and get a
+// notification once it's ready.
+#[derive(Default)]
+struct Waiter {
+    mutex: Mutex<bool>,
+    condvar: Condvar,
+}
+
+impl Waiter {
+    fn wait(&self) {
+        let mut ready = self.mutex.lock().unwrap();
+        while !*ready {
+            ready = self.condvar.wait(ready).unwrap();
+        }
+    }
+
+    fn notify(&self) {
+        let mut ready = self.mutex.lock().unwrap();
+        *ready = true;
+        self.condvar.notify_all();
+    }
+
+    fn finished(&self) -> bool {
+        *self.mutex.lock().unwrap()
+    }
+}
+
+// Used to track when an entry is being loaded. Once this is dropped,
+// we know the entry finished loading or failed to do so, and we can
+// unlock other clients waiting for it.
+pub struct ResourceLoadGuard<K> {
+    waiter: Arc<Waiter>,
+    key: K,
+}
+
+impl<K> Drop for ResourceLoadGuard<K> {
+    fn drop(&mut self) {
+        self.waiter.notify();
+    }
+}
+
+// Use to wait until an entry is ready to be used
+pub struct ResourceWaiter(Arc<Waiter>);
```

Comment:
This, and the `Waiter` could be removed and just let the consumer loop/sleep until the resource is ready. This just makes the wait more efficient by notifying when it's actually ready.